internal/resolver/memory.go

```diff
@@ -19,48 +19,93 @@ import (
 	"context"
 	"sync"
 
+	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/container/set"
 )
 
 // Memory is a memory based resolver.
 type Memory struct {
-	index sync.Map // map[string]ocispec.Descriptor
+	lock  sync.RWMutex
+	index map[string]ocispec.Descriptor
+	tags  map[digest.Digest]set.Set[string]
 }
 
 // NewMemory creates a new Memory resolver.
 func NewMemory() *Memory {
-	return &Memory{}
+	return &Memory{
+		index: make(map[string]ocispec.Descriptor),
+		tags:  make(map[digest.Digest]set.Set[string]),
+	}
 }
 
 // Resolve resolves a reference to a descriptor.
 func (m *Memory) Resolve(_ context.Context, reference string) (ocispec.Descriptor, error) {
-	desc, ok := m.index.Load(reference)
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	desc, ok := m.index[reference]
 	if !ok {
 		return ocispec.Descriptor{}, errdef.ErrNotFound
 	}
-	return desc.(ocispec.Descriptor), nil
+	return desc, nil
 }
 
 // Tag tags a descriptor with a reference string.
 func (m *Memory) Tag(_ context.Context, desc ocispec.Descriptor, reference string) error {
-	m.index.Store(reference, desc)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.index[reference] = desc
+	tagSet, ok := m.tags[desc.Digest]
+	if !ok {
+		tagSet = set.New[string]()
+		m.tags[desc.Digest] = tagSet
+	}
+	tagSet.Add(reference)
 	return nil
 }
 
 // Untag removes a reference from index map.
 func (m *Memory) Untag(reference string) {
-	m.index.Delete(reference)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	desc, ok := m.index[reference]
+	if !ok {
+		return
+	}
+	delete(m.index, reference)
+	tagSet := m.tags[desc.Digest]
+	tagSet.Delete(reference)
+	if len(tagSet) == 0 {
+		delete(m.tags, desc.Digest)
+	}
 }
 
 // Map dumps the memory into a built-in map structure.
-// Like other operations, calling Map() is go-routine safe. However, it does not
-// necessarily correspond to any consistent snapshot of the storage contents.
+// Like other operations, calling Map() is go-routine safe.
 func (m *Memory) Map() map[string]ocispec.Descriptor {
-	res := make(map[string]ocispec.Descriptor)
-	m.index.Range(func(key, value interface{}) bool {
-		res[key.(string)] = value.(ocispec.Descriptor)
-		return true
-	})
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	res := make(map[string]ocispec.Descriptor, len(m.index))
+	for key, value := range m.index {
+		res[key] = value
+	}
+	return res
```

Comment:
FYI: [maps.Clone](https://pkg.go.dev/maps#Clone) is introduced since `go 1.21`. Based on the `oras-go` support window, we can start using `maps.Clone` after the `go 1.22` release.

Reply:
/cc @Wwwsylvia